src/ReinforcementLearningCore/test/policies/agent.jl

```diff
@@ -0,0 +1,93 @@
+using ReinforcementLearningBase, ReinforcementLearningEnvironments
+using ReinforcementLearningCore: SRT, update!
+using ReinforcementLearningCore
+
+@testset "agent.jl" begin
+    @testset "Agent Cache struct" begin
+        srt = SRT{Int64, Float64, Bool}()
+        update!(srt, 2)
+        @test srt.state == 2
+        update!(srt, 1.0, true)
+        @test srt.reward == 1.0
+        @test srt.terminal == true
+    end
+
+    @testset "Trajectory SART struct compatibility" begin
+        srt_1 = SRT()
+        srt_2 = SRT{Any, Nothing, Nothing}()
+        srt_2.state = 1
+        srt_3 = SRT{Any, Any, Bool}()
+        srt_3.state = 1
+        srt_3.reward = 1.0
+        srt_3.terminal = true
+        trajectory = Trajectory(
+            CircularArraySARTTraces(; capacity = 1_000, reward=Float64=>()),
+            DummySampler(),
+        )
+        
+        @test_throws ArgumentError push!(trajectory, srt_1)
+        push!(trajectory, srt_2, 1)
+        @test length(trajectory.container) == 0
+        push!(trajectory, srt_3, 2)
+        @test length(trajectory.container) == 1
+        @test trajectory.container[:action] == [1]
+        push!(trajectory, srt_3, 3)
+        @test trajectory.container[:action] == [1, 2]
+        @test trajectory.container[:state] == [1, 1]
+    end
+
+    @testset "Agent Tests" begin
+        a_1 = Agent(
```

Comment:
When constructing an agent without explicitly supplying the cache struct, it will make one that has `Union{Any,Nothing} fields. This remains a perfomance problem I believe. Could we not insert in the `Agent(policy, trajectory)` constructor a routine that infers the type of each Trace based on the trajectory, and construct a typed cache. This information was already supplied by the user when constructing the trajectory. So it exists somewhere.

Reply:
So far, I haven't seen a meaningful performance hit for not specifying the cache, so I left it the way it is.

Reply:
If you say so then ok. It certainly won't be worse than before.